### PR TITLE
actors, functions, etc in parser

### PR DESCRIPTION
**Pull request by aethos**
… still need to make sure that pattern matching and lambdas work

### CODE REVIEW

**Comment by aethos** [comment]
we should merge this

File: src/ast.mli [hunk 1]
```diff
@@ -76,18 +70,28 @@ and stmt =
   | Break
   | Continue
 
-type func = {
+type pattern = {
```
Comment (Howon) [inline src/ast.mli]: what is this?

Comment (aethos) [reply src/ast.mli]: unfinished code

**Comment by aethos** [comment]
ok this is ready to merge i think

**Comment by Howon** [comment]
Let's make an output directory to put parsed_text.txt

File: src/ast.mli [hunk 2]
```diff
@@ -76,18 +70,37 @@ and stmt =
...
-type func = {
-  name : string;
-  return_t : types;
-  formals : formal list;
-  body : stmt list;
+
+type message = {
+    name: string;
+    formals: formal list;
+}
```
Comment (Howon) [inline src/ast.mli:77]: what happened to `sender` reference?

File: src/ast.mli [hunk 3]
```diff
@@ -76,18 +70,37 @@ and stmt =
...
-  formals : formal list;
-  receive: message_type list;
+  name: string;
+  formals: formal list;
+  body: stmt list;
+  functions: func list;
+  receive: pattern list;
+}
+
+type func = {
```
Comment (Howon) [inline src/ast.mli:93]: I feel like lambda can be parsed as just a nameless function?

File: src/parser.mly [hunk 4]
```diff
@@ -23,9 +23,8 @@
 %token EOF
 
 %nonassoc NOELSE
-%nonassoc PATTERN
 %right ASSIGN
-%left FUNC_ARG_TYPE ARROW FUN_RET_TYPE
+%left FUNC_ARG_TYPE ARROW FUNC_RET_TYPE
```
Comment (Howon) [inline src/parser.mly:27]: no fun :(

File: src/parser.mly [hunk 5]
```diff
@@ -61,7 +60,7 @@ message_list:
   | message_list message_decl { $2::$1 }
 
 message_decl:
-  TYPE_MESSAGE ID LPAREN formals_opt RPAREN         { None }
+  TYPE_MESSAGE ID LPAREN formals_opt RPAREN   { { name = $2; formals = $4 } }
```
Comment (Howon) [inline src/parser.mly:63]: same as above for `sender`

File: src/ast.mli (hunk 2 above)
Comment (vlad-scherbich) [reply src/ast.mli:77]: I think we discussed this with @aethos last weekend. But now I cannot recall why we thought it's OK to remove this.

`sender` can just be a `pid` right? Do you remember what we arrived at @aethos ?

Comment (aethos) [reply src/ast.mli:77]: i don't think we need a sender reference

File: src/parser.mly (hunk 4 above)
Comment (aethos) [reply src/parser.mly:27]: welcome 2 plt

File: src/ast.mli (hunk 2 above)
Comment (vlad-scherbich) [reply src/ast.mli:77]: per offline convo, `sender` cannot be determined at compile-time, so it won't be included in the `message` type.

File: src/ast.mli [hunk 6]
```diff
@@ -1,4 +1,4 @@
-type program = None
+type program = message list * actor list * func list
```
Comment (vlad-scherbich) [inline src/ast.mli:1]: this seems right, but why don't we have a `string list` tuple member in this definition?

File: src/parser.mly (hunk 4 above)
Comment (vlad-scherbich) [reply src/parser.mly:27]: lol 👍

File: src/ast.mli [hunk 7]
```diff
@@ -76,18 +70,37 @@ and stmt =
...
+  name: string;
+  return_t: types;
+  formals: formal list;
+  body: stmt list;
+}
+
+type lambda = {
+  formals: formal list;
+  return_t: types;
+  return_expr: expr;
```
Comment (vlad-scherbich) [inline src/ast.mli]: I wonder why `lambda` has `return_expr: expr;` as opposed to `body: stmt list;` as in `func`.
Isn't lambda basically a nameless function?

File: src/ast.mli (hunk 3 above)
Comment (vlad-scherbich) [reply src/ast.mli:93]: 👍

File: src/ast.mli (hunk 6 above)
Comment (aethos) [reply src/ast.mli:1]: a program is a list of messages, actors, and functions. Where do strings come into play?

File: src/ast.mli (hunk 2 above)
Comment (aethos) [reply src/ast.mli:77]: 👍

File: src/ast.mli (hunk 6 above)
Comment (vlad-scherbich) [reply src/ast.mli:1]: Just compared this with MicroC:
`type program = string list * func_decl list`

it's also a list of functions ... but seems you need string list for statements, expr and such?
https://github.com/mychrisdangelo/microc-withcomments/blob/master/ast.ml (edited)

**Comment by aethos** [comment]
merge ittttttt

**Review by vlad-scherbich — APPROVED** [review]

**Pull request merged**